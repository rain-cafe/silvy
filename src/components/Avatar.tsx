import styles from './Avatar.module.scss';
import Image from 'next/image';

type AvatarProps = {
    src: string;
}

export default function Avatar({
    src
}: AvatarProps) {
    return (
        <Image
            className={styles.avatar}
            width={40}
            height={40}
            src={src}
        />
    )
}