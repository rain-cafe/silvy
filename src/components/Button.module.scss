.button {
    display: inline-flex;
    align-items: center;
    border-radius: 5px;
    border: none;
    cursor: pointer;
    color: inherit;

    background: transparent;
    transition: 0.15s ease-in-out;
    transition-property: background-color, color;

    &:active {
        color: inherit;
    }

    &.sz-small {
        padding: 10px;
        font-size: 16px;
    }

    &.sz-medium {
        height: 50px;
        padding: 0 10px;
        font-size: 24px;
    }

    &.sz-large {
        height: 50px;
        padding: 0 20px;
        font-size: 30px;
        font-weight: bold;
    }

    &:hover {
        &.an-darken {
            background-color: rgba(0, 0, 0, 0.25);
        }

        &.an-swap {
            // Figure out if there's a way of clipping the text to be transparent
            background-color: #f492a5;
            color: #161b21;
        }
    }
}