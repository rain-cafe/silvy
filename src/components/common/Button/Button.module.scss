.button {
    transition: 0.15s ease-in-out;
    transition-property: background-color, color;

    &:hover {
        &.an-darken {
            background-color: rgba(0, 0, 0, 0.25);
        }

        &.an-swap {
            // Figure out if there's a way of clipping the text to be transparent
            background-color: #f492a5;
            color: #161b21;
        }
    }
}