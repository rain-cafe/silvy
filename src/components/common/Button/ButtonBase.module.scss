.buttonBase {
    display: inline-flex;
    gap: 10px;
    align-items: center;
    border-radius: 5px;
    border: none;
    cursor: pointer;
    color: inherit;
    min-height: 50px;
    user-select: none;

    background: transparent;

    &.sz-small {
        padding: 10px;
        font-size: 16px;
    }

    &.sz-medium {
        height: 50px;
        padding: 0 10px;
        font-size: 24px;
    }

    &.sz-large {
        height: 50px;
        padding: 0 20px;
        font-size: 30px;
        font-weight: bold;
    }

    &:active {
        color: inherit;
    }

    &.disabled {
        pointer-events: none;
    }
}