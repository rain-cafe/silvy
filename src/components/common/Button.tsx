import classnames from 'classnames';
import type { PolymorphicPropsWithoutRef } from 'react-polymorphic-types';
import styles from './Button.module.scss';
import { Animations, Sizes } from '../../constants/components';
import Link from 'next/link';

export const DEFAULT_ELEMENT = 'button';

export interface ButtonOwnProps {
    children: React.ReactNode;
    className?: string;
    size?: Sizes;
    animation?: Animations;
    disabled?: boolean;
}

export type ButtonProps<
    T extends React.ElementType = typeof DEFAULT_ELEMENT
> = PolymorphicPropsWithoutRef<ButtonOwnProps, T>;

export function Button<E extends React.ElementType = typeof DEFAULT_ELEMENT>({
    as,
    children,
    className,
    size = Sizes.SMALL,
    animation = Animations.DARKEN,
    disabled = false,
    ...extraProps
}: ButtonProps<E>) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const Element: React.ElementType = as || DEFAULT_ELEMENT;
    const props = {
        ...extraProps,
        className: classnames(
            styles.button,
            styles[`sz-${size}`],
            styles[`an-${animation}`],
            disabled && styles.disabled,
            className
        )
    };

    if (Element === Link) {
        return (
            <Element {...props}>
                <a {...props}>
                    {children}
                </a>
            </Element>
        );
    }

    return (
        <Element {...props}>
            {children}
        </Element>
    );
}