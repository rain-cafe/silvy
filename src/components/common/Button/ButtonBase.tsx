import type { PolymorphicPropsWithoutRef } from 'react-polymorphic-types';
import classNames from 'classnames';
import Link from 'next/link';
import { useReadOnlyCachedState } from '../../hooks/use-cached-state';
import styles from './ButtonBase.module.scss';
import React from 'react';
import { Sizes } from '../../../constants/components';

export const DEFAULT_ELEMENT = 'button';

export interface ButtonBaseOwnProps {
    children: React.ReactNode;
    className?: string;
    disabled?: boolean;
    size?: Sizes;
}

export type ButtonBaseProps<
    T extends React.ElementType = typeof DEFAULT_ELEMENT
> = PolymorphicPropsWithoutRef<ButtonBaseOwnProps, T>;

export function ButtonBase<E extends React.ElementType = typeof DEFAULT_ELEMENT>({
    as,
    className: externalClassName,
    children,
    disabled = false,
    size = Sizes.SMALL,
    ...extraProps
}: ButtonBaseProps<E>) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const Element: React.ElementType = as || DEFAULT_ELEMENT;
    const className = useReadOnlyCachedState(() => {
        return classNames(
            styles.buttonBase,
            styles[`sz-${size}`],
            externalClassName,
            {
                [styles.disabled]: disabled
            }
        );
    }, [disabled, externalClassName]);

    if (Element === Link) {
        return (
            <Element
                {...extraProps}
                className={className}
                disabled={disabled}
                draggable={false}
            >
                <a
                    data-testid='button-base'
                    className={className}
                >
                    {children}
                </a>
            </Element>
        );
    }

    return (
        <Element
            data-testid='button-base'
            {...extraProps}
            className={className}
            disabled={disabled}
            draggable={false}
        >
            {children}
        </Element>
    );
}