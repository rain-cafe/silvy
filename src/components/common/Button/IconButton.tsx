import { IconType } from 'react-icons';
import type { PolymorphicPropsWithoutRef } from 'react-polymorphic-types';
import classNames from 'classnames';
import {Button, ButtonOwnProps} from './Button';
import styles from './IconButton.module.scss';
import { Animations } from '../../../constants/components';
import { useReadOnlyCachedState } from '../../hooks/use-cached-state';

export interface IconButtonOwnProps extends Omit<ButtonOwnProps, 'children'> {
    icon: IconType;
    hoverIcon?: IconType;
    children?: React.ReactNode;
    className?: string;
    disabled?: boolean;
};

export type IconButtonProps<
    T extends React.ElementType
> = PolymorphicPropsWithoutRef<IconButtonOwnProps, T>;

export function IconButton<E extends React.ElementType>({
    as,
    icon: Icon,
    hoverIcon: HoverIcon,
    children,
    className: externalClassName,
    ...extraProps
}: IconButtonProps<E>) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const Element: React.ElementType = as;
    const className = useReadOnlyCachedState(() => {
        return classNames(
            styles.iconButton,
            externalClassName,
            {
                [styles.hasHoverIcon]: HoverIcon
            }
        );
    }, [HoverIcon, externalClassName]);

    return (
        <Button
            data-testid='icon-button'
            {...extraProps}
            as={Element}
            animation={HoverIcon ? Animations.NONE : Animations.DARKEN}
            className={className}
        >
            <Icon className={styles.icon} size={30} />
            {HoverIcon && (
                <HoverIcon className={styles.hoverIcon} size={30} />
            )}
            {children}
        </Button>
    );
}