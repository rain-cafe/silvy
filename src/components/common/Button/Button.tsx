import classnames from 'classnames';
import { PolymorphicPropsWithoutRef } from 'react-polymorphic-types';
import styles from './Button.module.scss';
import { ButtonBase, ButtonBaseOwnProps } from './ButtonBase';
import { useReadOnlyCachedState } from '../../hooks/use-cached-state';
import { Animations } from '../../../constants/components';

export interface ButtonOwnProps extends ButtonBaseOwnProps {
    animation?: Animations;
}

export type ButtonProps<
    T extends React.ElementType
> = PolymorphicPropsWithoutRef<ButtonOwnProps, T>;

export function Button<E extends React.ElementType>({
    as,
    className: externalClassName,
    animation = Animations.DARKEN,
    ...extraProps
}: ButtonProps<E>) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const Element: React.ElementType = as;
    const className = useReadOnlyCachedState(() => {
        return classnames(
            styles.button,
            styles[`an-${animation}`],
            externalClassName
        );
    }, [animation, externalClassName]);

    return (
        <ButtonBase
            as={Element}
            className={className}
            data-testid='button'
            {...extraProps}
        />
    );
}