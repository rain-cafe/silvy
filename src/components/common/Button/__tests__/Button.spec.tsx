import React from 'react';
import { createRenderComponent } from '@rain/test-utils/render';
import { chance } from '@rain/test-utils/chance';
import { Animations, Sizes } from '../../../../constants/components';
import { Button, ButtonProps } from '../Button';
import styles from '../Button.module.scss';

describe('Component(Button)', () => {
    const renderButton = createRenderComponent<ButtonProps<React.ElementType>>(Button, {
        children: 'UwU'
    });

    it('should display the Button', () => {
        const { baseElement } = renderButton()

        expect(baseElement).toMatchSnapshot()
    });

    it('should support overriding the animation', () => {
        const { getByTestId, expectedProps } = renderButton({
            animation: Animations.SWAP
        });

        expect(getByTestId('button')).toHaveClass(styles[`an-${expectedProps.animation}`]);
    });

    it('should support providing a class', () => {
        const { getByTestId, expectedProps } = renderButton({
            className: chance.string({ alpha: true })
        });

        expect(getByTestId('button')).toHaveClass(expectedProps.className);
    });
});
