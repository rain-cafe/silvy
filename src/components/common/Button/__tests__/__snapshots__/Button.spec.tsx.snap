// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component(Button) should display the Button 1`] = `
<body>
  <div>
    <button
      class="buttonBase sz-small button an-darken"
      data-testid="button"
      draggable="false"
    >
      UwU
    </button>
  </div>
</body>
`;
